storagemarket/types.go

```diff
@@ -108,6 +114,115 @@ type MinerDeal struct {
 	SectorNumber      abi.SectorNumber
 }
 
+// NewDealStages creates a new DealStages object ready to be used.
+// EXPERIMENTAL; subject to change.
+func NewDealStages() *DealStages {
+	return &DealStages{}
+}
+
+// DealStages captures a timeline of the progress of a deal, grouped by stages.
+// EXPERIMENTAL; subject to change.
+type DealStages struct {
+	// Stages contains an entry for every stage that the deal has gone through.
+	// Each stage then contains logs.
+	Stages []*DealStage
+}
+
+// DealStages captures data about the execution of a deal stage.
+// EXPERIMENTAL; subject to change.
+type DealStage struct {
+	// Human-readable fields.
+	// TODO: these _will_ need to be converted to canonical representations, so
+	//  they are machine readable.
+	Name             string
+	Description      string
+	ExpectedDuration string
+
+	// Timestamps.
+	// TODO: may be worth adding an exit timestamp. It _could_ be inferred from
+	//  the start of the next stage, or from the timestamp of the last log line
+	//  if this is a terminal stage. But that's non-determistic and it relies on
+	//  assumptions.
+	CreatedTime cbg.CborTime
+	UpdatedTime cbg.CborTime
+
+	// Logs contains a detailed timeline of events that occurred inside
+	// this stage.
+	Logs []*Log
+}
+
+// Log represents a point-in-time event that occurred inside a deal stage.
+// EXPERIMENTAL; subject to change.
+type Log struct {
+	// Log is a human readable message.
+	// TODO: this _may_ need to be converted to a canonical data model so it
+	//  is machine-readable.
+	Log         string
+	UpdatedTime cbg.CborTime
+}
+
+// GetStage returns the DealStage object for a named stage, or nil if not found.
+// TODO: the input should be a strongly-typed enum instead of a free-form string.
+// TODO: drop Get from GetStage to make this code more idiomatic. Return a
+//  second ok boolean to make it even more idiomatic.
+// EXPERIMENTAL; subject to change.
+func (ds *DealStages) GetStage(stage string) *DealStage {
+	if ds == nil {
+		return nil
+	}
+
+	for _, s := range ds.Stages {
+		if s.Name == stage {
+			return s
+		}
+	}
+
+	return nil
+}
+
+// AddStageLog adds a log to the specified stage, creating the stage if it
+// doesn't exist yet.
+// EXPERIMENTAL; subject to change.
+func (ds *DealStages) AddStageLog(stage, description, expectedDuration, msg string) {
+	if ds == nil {
+		return
```

Comment:
We don't want to create a DealStages instance here? I guess it would allow us to display partial logs for deals that were in progress before this code got deployed?

```go
if ds == nil {
    *ds = DealStages{}
}
```

Reply:
Actually, this is not possible. We would need to return the new `DealStages` object, similar to how hashicorp/go-multierror.Append works: https://pkg.go.dev/github.com/hashicorp/go-multierror#Append

Reply:
This is done for backward compatibility with deals that are created with earlier versions.